### PR TITLE
WIP: First cut on clipped factorization analysis

### PR DESCRIPTION
**Pull request by timodonnell**
Another WIP on a new analysis method. Not ready for merging but feedback welcome.

### CODE REVIEW

File: phip/clipped_factorization_analysis.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+import numpy
+
+import tensorflow as tf
+from tensorflow.contrib.distributions import percentile
+
+
+def do_clipped_factorization_analysis(
+        counts_df,
+        rank=3,
+        truncate_percentile=99.9,
```
Comment (laserson) [inline phip/clipped_factorization_analysis.py]: How can we determine what the best value for this is?

File: phip/clipped_factorization_analysis.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    # Variables
+    a = tf.Variable(
+        numpy.random.rand(n, rank), name="A", dtype="float32")
+    b = tf.Variable(
+        numpy.random.rand(rank, s), name="B", dtype="float32")
+    truncate_threshold = tf.Variable(reads_per_million.max().max())
+
+    # Derived quantities
+    reconstruction = tf.matmul(
+        tf.gather(a, minibatch_indices), b)
```
Comment (laserson) [inline phip/clipped_factorization_analysis.py]: Is there an implication for sampling rows but not cols? Or is this basically just a pragmatic issue?

File: phip/clipped_factorization_analysis.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    train_step = tf.train.AdamOptimizer(
+        learning_rate=learning_rate).minimize(loss)
+
+    init = tf.global_variables_initializer()
+
+    best_cost_value = None
+    last_log_at = 0
+    with tf.Session() as session:
+        session.run(init)
+        all_indices = numpy.array(list(range(reads_per_million.shape[0])))
```
Comment (laserson) [inline phip/clipped_factorization_analysis.py]: `all_indices = numpy.arange(reads_per_million.shape[0])`?

File: phip/clipped_factorization_analysis.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
+import time
+
+import pandas
```
Comment (laserson) [inline phip/clipped_factorization_analysis.py]: I'd go with either

```
import pandas as pd
```

or

```
from pandas import DataFrame
```

File: phip/clipped_factorization_analysis.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+import numpy
```
Comment (laserson) [inline phip/clipped_factorization_analysis.py]: same...I guess `import numpy as np`

File: phip/clipped_factorization_analysis.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+    # unclipped_term is based only on the minimum unclipped error for each
+    # clone. The intuition is that we know for every clone at least one sample
+    # must be a non-hit (e.g. a beads only sample), and so should be well modeled
+    # by the background process.
+    unclipped_term = tf.reduce_min(
+        tf.pow(differences, 2), axis=1)
+    loss = (
+        tf.reduce_mean(
+            tf.pow(tf.minimum(differences, truncate_threshold), 2))
```
Comment (laserson) [inline phip/clipped_factorization_analysis.py]: Strictly speaking, is this the Frobenius norm? Don't you need to sum the matrix elts and take the square root? This would change the values of the gradients, no?

File: phip/clipped_factorization_analysis.py (hunk 1 above)
Comment (timodonnell) [reply phip/clipped_factorization_analysis.py]: We should be pretty robust to this choice (e.g. 99 vs. 99.9 vs. 99.99). It just needs to be a lower-bound on fraction of non-hits. May be good to try different values though as we analyze the next few data sets

File: phip/clipped_factorization_analysis.py (hunk 2 above)
Comment (timodonnell) [reply phip/clipped_factorization_analysis.py]: Yeah taking the minibatches row-wise is really just a pragmatic thing, since that's the long dimension in our data

File: phip/clipped_factorization_analysis.py (hunk 3 above)
Comment (timodonnell) [reply phip/clipped_factorization_analysis.py]: 👍

File: phip/clipped_factorization_analysis.py (hunk 4 above)
Comment (timodonnell) [reply phip/clipped_factorization_analysis.py]: 👍

File: phip/clipped_factorization_analysis.py (hunk 5 above)
Comment (timodonnell) [reply phip/clipped_factorization_analysis.py]: 👍

File: phip/clipped_factorization_analysis.py (hunk 6 above)
Comment (timodonnell) [reply phip/clipped_factorization_analysis.py]: You're correct, but while it will change the value of the gradients it will not change the coordinates of the minimum point, so we are still optimizing the advertised function. Taking the mean instead of sum is mean to keep the loss roughly on a similar scale to the data entries (so e.g. learning rate isn't tightly dependent on matrix size) and skipping the square root is for speed.

**Comment by timodonnell** [comment]
Thanks for the comments @laserson ! Updated to address comments and rebase onto current master. After playing with this for a while I think it's reasonable to merge after we address any remaining things you notice. Additional changes I'm making now:

* Added unit test
* Rename all occurrences of "truncated factorization" to "clipped factorization"
* Better docstring
* Small tweaks to string formatting so as not to require python 3.6

**Comment by laserson** [comment]
lgtm! feel free to merge, but I'd do a rebase onto master beforehand just in case.
